sql/src/main/java/io/crate/planner/operators/LogicalPlan.java

```diff
@@ -0,0 +1,117 @@
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * LogicalPlan is a tree of "Operators"
+ * This is a representation of the logical order of operators that need to be executed to produce a correct result.
+ *
+ * {@link #build(Planner.Context, ProjectionBuilder, int, int, OrderBy)} is used to create the
+ * actual "physical" execution plan.
+ *
+ * A Operator is something like Limit, OrderBy, HashAggregate, Join, Union, Collect
+ * <pre>
+ *     select x, y, z from t1 where x = 10 order by x limit 10:
+ *
+ *     Limit 10
+ *        |
+ *     Order By x
+ *         |
+ *     Collect [x, y, z]
+ * </pre>
+ *
+ * {@link #build(Planner.Context, ProjectionBuilder, int, int, OrderBy)} is called on the "root" and flows down.
+ * Each time each operator may provide "hints" to the children so that they can decide to eagerly apply parts of the
+ * operations
+ *
+ * This allows us to create execution plans as follows::
+ *
+ * <pre>
+ *     select x, y, z from t1 where order by x limit 10;
+ *
+ *
+ *          Merge
+ *         /  limit10
+ *       /         \
+ *     Collect     Collect
+ *     limit 10    limit 10
+ *
+ * </pre>
+ */
+public interface LogicalPlan {
+
+    interface Builder {
+
+        /**
+         * Create a LogicalPlan node
+         *
+         * @param usedColumns The columns the "parent" is using.
+         *                    This is used to create plans which utilize query-then-fetch.
+         *                    For example:
+         *                    <pre>
+         *                       select a, b, c from t1 order by a limit 10
+         *
+         *                       EvalFetch (usedColumns: [a, b, c])
+         *                         outputs: [a, b, c]   (b, c resolved using _fetch)
+         *                         |
+         *                       Limit 10 (usedColumns: [])
+         *                         |
+         *                       Order (usedColumns: [a]
+         *                         |
+         *                       Collect (usedColumns: [a] - inherited from Order)
+         *                         outputs: [_fetch, a]
+         *                    </pre>
+         */
+        LogicalPlan build(Set<Symbol> usedColumns);
+    }
+
+    /**
+     * Uses the current shard allocation information to create a physical execution plan.
+     * <br />
+     * {@code limit}, {@code offset}, {@code order} can be passed from one operator to another. Depending on the
+     * operators implementation. Operator may choose to make use of this information, but can also ignore it.
+     */
+    Plan build(Planner.Context plannerContext,
+               ProjectionBuilder projectionBuilder,
+               int limit,
+               int offset,
+               @Nullable OrderBy order);
+
+    /**
+     * Used to generate optimized operators.
+     * E.g. Aggregate(count(*)) + Collect -> Count
+     */
+    LogicalPlan tryCollapse();
```

Comment:
Default implementation should be `return this` maybe? I can also see that we want to force implementing this to eventually collapse the tree.

Reply:
Thought about that, but as you pointed out I wanted to force the implementation so that the tree is collapsed.